File: test-suite/js/security/reverse-proxy-ssrf-clean.ts
import { createProxyMiddleware } from "http-proxy-middleware";
import httpProxy from "http-proxy";
import { NextResponse } from "next/server";

type ExpressRequest = {
  query: Record<string, string | undefined>;
  headers: Record<string, string | undefined>;
};

type ExpressResponse = {
  end(): void;
};

const allowedProxyHosts = new Set(["api.example.com", "static.example.com"]);
const proxy = httpProxy.createProxyServer({});

function validateProxyTarget(raw: string | null | undefined): string {
  const parsed = new URL(raw ?? "https://api.example.com/health");
  if (parsed.protocol !== "https:" || !allowedProxyHosts.has(parsed.hostname)) {
    throw new Error("blocked proxy target");
  }
  return parsed.toString();
}

function isAllowedProxyTarget(raw: string | undefined): boolean {
  if (!raw) {
    return false;
  }
  const parsed = new URL(raw);
  return parsed.protocol === "https:" && allowedProxyHosts.has(parsed.hostname);
}

export function proxyMiddlewareFromQuery(req: ExpressRequest) {
  const target = validateProxyTarget(req.query.upstream);
  return createProxyMiddleware({
    target,
    changeOrigin: true,
  });
}

export function proxyWebFromHeader(req: ExpressRequest, res: ExpressResponse) {
  const upstreamUrl = req.headers["x-upstream"];
  if (!isAllowedProxyTarget(upstreamUrl)) {
    throw new Error("blocked proxy target");
  }
  proxy.web(req, res, {
    target: upstreamUrl,
  });
}

export function proxyRouterFromHeader() {
  return createProxyMiddleware({
    router: (req: ExpressRequest) => validateProxyTarget(req.headers["x-proxy-target"]),
  });
}

export function nextRewriteProxy(request: Request) {
  const rewriteUrl = validateProxyTarget(new URL(request.url).searchParams.get("proxy"));
  return NextResponse.rewrite(rewriteUrl);
}

File: test-suite/golang/security/reverse_proxy_ssrf_clean.go
package security

import (
	"errors"
	"net/http"
	"net/http/httputil"
	"net/url"
)

var allowedProxyHosts = map[string]bool{
	"api.example.com": true,
	"cdn.example.com": true,
}

func safeProxyURL(raw string) (*url.URL, error) {
	upstream, err := url.Parse(raw)
	if err != nil {
		return nil, err
	}
	if upstream.Scheme != "https" || !allowedProxyHosts[upstream.Hostname()] {
		return nil, errors.New("blocked proxy target")
	}
	return upstream, nil
}

func proxyWithSafeTarget(r *http.Request) (*httputil.ReverseProxy, error) {
	upstream, err := safeProxyURL(r.URL.Query().Get("target"))
	if err != nil {
		return nil, err
	}
	return httputil.NewSingleHostReverseProxy(upstream), nil
}

func rewriteWithSafeTarget(r *http.Request) (*httputil.ReverseProxy, error) {
	upstream, err := safeProxyURL(r.Header.Get("X-Upstream"))
	if err != nil {
		return nil, err
	}
	return &httputil.ReverseProxy{
		Rewrite: func(pr *httputil.ProxyRequest) {
			pr.SetURL(upstream)
		},
	}, nil
}

func directorWithSafeTarget(r *http.Request) (*httputil.ReverseProxy, error) {
	upstream, err := safeProxyURL("https://" + r.Host)
	if err != nil {
		return nil, err
	}
	return &httputil.ReverseProxy{
		Director: func(out *http.Request) {
			out.URL.Scheme = upstream.Scheme
			out.URL.Host = upstream.Host
		},
	}, nil
}
